core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -20,33 +20,143 @@
+  private static final RemovalListener<TableIdentifier, Table> identLoggingRemovalListener =
+      (key, value, cause) -> LOG.debug("{} was evicted from the TableCache with {} cause", key, cause);
+
   public static Catalog wrap(Catalog catalog) {
-    return wrap(catalog, true);
+    return wrap(catalog, false, 0);
+  }
+
+  public static Catalog wrap(Catalog catalog, boolean expirationEnabled, long expirationIntervalMilllis) {
+    return wrap(catalog, true, expirationEnabled, expirationIntervalMilllis);
+  }
+
+  public static Catalog wrap(
+      Catalog catalog, boolean caseSensitive, boolean expirationEnabled, long expirationIntervalMillis) {
+    return new CachingCatalog(catalog, caseSensitive, expirationEnabled, expirationIntervalMillis);
   }
 
-  public static Catalog wrap(Catalog catalog, boolean caseSensitive) {
-    return new CachingCatalog(catalog, caseSensitive);
+  @VisibleForTesting
+  static Catalog wrap(Catalog catalog, boolean expirationEnabled, long expirationIntervalMillis, Ticker ticker) {
+    return new CachingCatalog(catalog, true, expirationEnabled, expirationIntervalMillis, ticker);
   }
 
-  private final Cache<TableIdentifier, Table> tableCache = Caffeine.newBuilder().softValues().build();
   private final Catalog catalog;
   private final boolean caseSensitive;
+  private final boolean expirationEnabled;
+  private final long expirationIntervalMillis;
+  private final Cache<TableIdentifier, Table> tableCache;
+
+  private CachingCatalog(Catalog catalog,
+                         boolean caseSensitive,
+                         boolean isExpirationEnabled,
+                         long expirationIntervalInMillis) {
+    this(catalog, caseSensitive, isExpirationEnabled, expirationIntervalInMillis, Ticker.systemTicker());
+  }
 
-  private CachingCatalog(Catalog catalog, boolean caseSensitive) {
+  private CachingCatalog(Catalog catalog,
+                         boolean caseSensitive,
+                         boolean isExpirationEnabled,
+                         long expirationIntervalMillis,
+                         Ticker ticker) {
     this.catalog = catalog;
     this.caseSensitive = caseSensitive;
+    this.expirationEnabled = isExpirationEnabled;
+    this.expirationIntervalMillis = expirationIntervalMillis;
+
+    this.tableCache = createTableCache(ticker);
+  }
+
+  /**
+   * Return the age of an entry in the cache.
+   * <p>
+   * This method is only visible for testing the cache expiration policy, as cache invalidation is handled
+   * by the catalog and not the cache itself.
+   * <p>
+   * Returns the age of the cache entry corresponding to the identifier,  or {@code Optional.empty} if the table
+   * is not present in the cache or if no expireAfterAccess policy is present in this CachingCatalog.
+   */
+  @VisibleForTesting
+  Optional<Duration> cachedEntryAge(TableIdentifier identifier) {
+    return tableCache.policy()
```

Comment:
maybe `tableCache.policy().expireAfterAccess().get().ageOf(identifier)`?

Reply:
Yeah if this is just a testing method then it doesn't need the overhead of the option.

Reply:
This was moved into `TestableCachingCatalog`. The Option is still kept, only because that's what `ageOf` returns.

I can move the `.get` call into it, but that's why `Optional<Duration>` is still present in the `TestableCachingCatalog`.